in-progress/8757-fees/design.md

```diff
@@ -0,0 +1,192 @@
+- `minimum_proving_cost_per_mana` - the minimum cost in wei for proving a unit of mana
+- `maximum_proving_cost_multiplier_per_block` - the maximum percentage increase in the cost of proving a unit of mana per block
+- `minimum_fee_asset_price` - the minimum price in wei of the fee asset (e.g. TST)
+- `maximum_fee_asset_price_multiplier_per_block` - the maximum percentage increase in the price of the fee asset per block
+- `underlying_base_fee_oracle_update_interval` - the minimum number of slots between updates to the underlying base fee oracle
+- `congestion_factor_multiplier` - the constant factor to multiply the congestion factor by
+- `maximum_epoch_proof_quote_fee` - the maximum fee a prover can submit for an epoch proof quote
+
+## Oracles
+
+### `fee_asset_per_wei`
+
+This will be stored on the rollup contract. On proposing a block, proposers will be able to adjust the rate by up to `maximum_fee_asset_price_multiplier_per_block`.
+
+Has a minimum value of `minimum_fee_asset_price`.
+
+### `proving_cost_per_mana`
+
+This will be stored on the rollup contract. On proposing a block, proposers will be able to adjust the rate by up to `maximum_proving_cost_multiplier_per_block`.
+
+Has a minimum value of `minimum_proving_cost_per_mana`.
+
+### `wei_per_l1_gas` and `wei_per_l1_blob_gas`
+
+The rollup contract will maintain an `wei_per_l1_gas` oracle. 
+
+Same for `wei_per_l1_blob_gas`.
+
+They will be updated up to every `underlying_base_fee_oracle_update_interval` slots.
+
+
+## Getting the cost of a block in wei/mana
+
+When a proposer is building a block, it calculates the total L1 cost of the block in wei by summing:
+- `l1_gas_per_block_proposed` * `wei_per_l1_gas`
+- `blobs_per_block` * `GAS_PER_BLOB` * `wei_per_l1_blob_gas`
+- `l1_gas_per_epoch_verified` * `wei_per_l1_gas` / `l2_slots_per_l2_epoch`
+
+The cost of the block in wei/mana is then the sum of the following:
+- the total L1 cost in wei divided by `target_mana_per_block`
+- `proving_cost_per_mana`
+
+## Getting the Base Fee in wei/mana
+
+After simulating a block, the mana used by each transaction is known.
+
+A proposer can compute a block's `total_mana_used` by summing the `mana_used` fields of all transactions in the block.
+
+This `total_mana_used` will be part of the L2 block header.
+
+The rollup contract will maintain a `excess_mana` variable for the proven and pending chains.
+
+The `excess_mana` is computed from the previous block as the difference between `total_mana_used` and `target_mana_per_block`.
+
+Compute the L2 congestion factor as the "fake exponential" of the ratio of `total_mana_used` plus the previous block's `excess_mana` to `target_mana_per_block`, and multiply by a constant factor of 1e9:
+
+```math
+\text{congestion factor} = \text{congestion factor multiplier} * \text{exp}\left(\frac{\text{total mana used} + \text{previous excess mana}}{\text{target mana per block}}\right)
```

Comment:
- The thing you are calling the `congestion_factor_multiplier` is essentially the minimum congestion factor.
- It is not the target mana per block as the denominator, it is a `update_fraction` which can be computed from target mana but not directly. For example, if you want a max increment of 12%, you can use mana target divided by 0.117, e.g., if the total fraction is 0.117 that is an increment of 12% so for each entity total mana per block you are above it will add approx 12%.